### DISCUSSION TITLE
Dimming 1 channel instead of 5 (Sunricher 4 in 1 Universal ZigBee LED Controller - dim mode)

### DISCUSSION THREAD

**Discussion by Hubert-BUYSSENS** [Koenkk/zigbee2mqtt#24827] [Devices]
I've bought Sunricher ref. SR-ZG1029-5C to dim 5 different zone. 
This product have 4 dimming mode : DIM_5ch CCT RGBW RGBCCT.
manual : https://www.sunricher.com/media/resources/manual/SR-ZG1029-5C%20V2%20instruction.pdf

I’m quite a newbie with home assistant and Z2M.

The integration in Z2M is detected as another device, so we don't have full access to diming function :
- In RGB+CCT mode, it's detected as Illuminize 511.000 model as it uses the Zigbee model HK-ZD-RGBCCT-A. => only RGB is available.
- In DIM mode, it's detected it as Sunricher SRP-ZG9105-CC device (reports as Zigbee model HK-ZD-DIM-A, only one channel) => the 5 channels are dimmed together, as if it's only one channel.

I've find on this blog that martinweber achieve to create a configuration that resolve the problem on RGB+CCT mode :
https://cgrebel.com/2023/01/zigbee2mqtt-homeassistant-sunricher-sr-zg1029-5c-support/
And read also :
https://www.zigbee2mqtt.io/advanced/support-new-devices/01_support_new_devices.html#_2-1-extending-the-external-definition

So I've tried to create a similar configuration for 5 channel dimming :

```
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const extend = require('zigbee-herdsman-converters/lib/extend');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    zigbeeModel: ['HK-ZD-DIM-A'],
    // https://www.sunricher.com/dim-cct-rgbw-rgb-cct-4-in-1-constant-voltage-zigbee-led-controller-sr-zg1029-5c.html
    // The controller can be put into different modes using DIP switches: DIM, CCT, RGBW, and RGB+CCT
    // Also sold as ISOLed Model 114729: https://www.isoled.shop/de/zigbee-30-pwm-dimmer-fuer-led-flexbaenderspots-5-kanal-12-24v-dc-5x4a_114729.html
    model: 'SR-ZG1029-5C',
    vendor: 'Sunricher',
    whiteLabel: [{vendor: 'Sunricher', model: 'SR-ZG1029-5C'}],
    description: 'DIM CCT RGBW RGB+CCT 4-in-1 Constant Voltage Zigbee LED Controller',
    extend: extend.light_onoff_brightness(),
    exposes: [
        e.light_brightness().withEndpoint('l1'),
        e.light_brightness().withEndpoint('l2'),
        e.light_brightness().withEndpoint('l3'),
        e.light_brightness().withEndpoint('l4'),
        e.light_brightness().withEndpoint('l5')
    ],
    endpoint: (device) => {
        return {'l1' : 1, 'l2' : 2, 'l3' : 3, 'l4' : 5, 'l5' : 5 };
    },
    meta: {multiEndpoint: true},
};


module.exports = definition;
```
But log report this error :
z2m: Failed to load external converter file 'SR-ZG1029-5C-DIM.js' (Cannot find module 'zigbee-herdsman-converters/lib/extend'
Require stack:
- /app/dist/util/utils.js
- /app/dist/util/settings.js
- /app/index.js)

Then with some internet research (https://github.com/Koenkk/zigbee2mqtt/releases/tag/1.36.1 ), I’ve find that I have to remove the line:
const extend = require('zigbee-herdsman-converters/lib/extend')

But the error became : 
z2m: Failed to load external converter file ‘SR-ZG1029-5C-DIM.js’ (extend is not defined)
So it seems to need the lib/extend… but don’t know where, if it’s not in this external converter?

So I’ve tried to remove the error with removing what’s after extend:

```
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    zigbeeModel: ['HK-ZD-DIM-A'],
    // https://www.sunricher.com/dim-cct-rgbw-rgb-cct-4-in-1-constant-voltage-zigbee-led-controller-sr-zg1029-5c.html
    // The controller can be put into different modes using DIP switches: DIM, CCT, RGBW, and RGB+CCT
    // Also sold as ISOLed Model 114729: https://www.isoled.shop/de/zigbee-30-pwm-dimmer-fuer-led-flexbaenderspots-5-kanal-12-24v-dc-5x4a_114729.html
    model: 'SR-ZG1029-5C',
    vendor: 'Sunricher',
    whiteLabel: [{vendor: 'Sunricher', model: 'SR-ZG1029-5C'}],
    description: 'DIM CCT RGBW RGB+CCT 4-in-1 Constant Voltage Zigbee LED Controller',
    extend: [],
    exposes: [
        e.light_brightness().withEndpoint('l1'),
        e.light_brightness().withEndpoint('l2'),
        e.light_brightness().withEndpoint('l3'),
        e.light_brightness().withEndpoint('l4'),
        e.light_brightness().withEndpoint('l5')
    ],
    endpoint: (device) => {
        return {'l1' : 1, 'l2' : 2, 'l3' : 3, 'l4' : 5, 'l5' : 5 };
    },
    meta: {multiEndpoint: true},
};

module.exports = definition;
```

The external converter is loaded, HA show me 5 ON/OFF and brightness, but when I’m trying to move brightness HA report the error :
z2m: No converter available for 'brightness' (undefined)

I should miss something and need some help...

**Comment by Micha70** [comment]
Hi @Hubert-BUYSSENS, did you solved that problem? I'm searching for an similar solution....

**Reply by Hubert-BUYSSENS** [reply]
Hi @Micha70 , not yet, I have all my house to renovate, so I've postponned this problem... hopping that, in the meantime, it could be resolve with a Z2M update ;)
So I'm using it like a single zone for the moment.
